### DISCUSSION TITLE
Cannot run local installation of the Serverless Framework by the outdated global version.

### DISCUSSION THREAD

**Discussion by alexrygalov** [serverless/serverless#10595] [Q&A] [answered]
Hi there! 👋

Yesterday started receiving an error during sls deployment:
We use GItHub runners without docker - https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md
Checked recently updated software, nothing suspicious. Tried to use v3, it solves that particular problem but brings a new ones.
I'm just curious, what might changed with perfectly working v2.

```shell
Run `npm audit` for details.
Deploying dev API...
Deploying miscellaneous microservice...
Installing plugins into miscellaneous microservice
Serverless: Installing plugin "serverless-python-requirements@latest" (this might take a few seconds...)
Serverless: Successfully installed "serverless-python-requirements@latest"
Serverless: Installing plugin "serverless-sentry@latest" (this might take a few seconds...)
Serverless: Successfully installed "serverless-sentry@latest"
Serverless: Installing plugin "serverless-prune-plugin@latest" (this might take a few seconds...)
Serverless: Successfully installed "serverless-prune-plugin@latest"

changed 1 package, and audited 431 packages in 928ms

50 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
Plugins install successfully into miscellaneous microservice
Serverless: Running "serverless" installed locally (in service node_modules)
 
 Serverless Error ----------------------------------------
 
  Cannot run local installation of the Serverless Framework by the outdated global version.
  Please upgrade via:
  
  npm install -g serverless
  Note: Latest release can run any version of the locally installed Serverless Framework.
  
  Alternatively run locally installed version directly via:
  
  npx serverless <command> <options>
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information ---------------------------
     Operating System:          linux
     Node Version:              16.13.2
     Framework Version:         2.72.2
     Plugin Version:            5.5.4
     SDK Version:               4.3.0
     Components Version:        3.18.2
 
💩Deploy failed
```

The version is pinned in the GitHub **workflow.yaml**

```shell
    - name: Deploy
      if: ${{ env.ENVIRONMENT != '' }}
      shell: bash -l -eo pipefail {0}
      run: |
        export NVM_DIR="$HOME/.nvm"
        [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
        npm install -g serverless@2.72.2
```

And in the **serverless.yaml** config

```shell
# Version of the framework to use
frameworkVersion: ">=2.63.0 <2.72.2"
```

**Comment by pgrzesik** [comment, marked as answer]
Hello @alexrygalov - it looks like you might still have a locally installed v3 version (in project's `node_modules`). Could you try the same thing on fresh environment, without v3 installed?

**Reply by alexrygalov** [reply]
Thank's for the direction. Looks like it is some cache on GitHub Hosted Instances. Tried to run it on self-hosted instance in docker all good.
Also tried to install a node with nvm and the use installed version, result was the same.
Anyway it should be better run in docker.

**Reply by pgrzesik** [reply]
I'm glad you were able to resolve the problem 🙌

**Reply by StuartBertram** [reply]
How do we identify whether there's a locally installed v3? Is there a fix for this that works with GitHub Actions?

**Reply by pgrzesik** [reply]
You would have to install the v3 version locally first for it to be available - what problem are you facing?

**Reply by StuartBertram** [reply]
We've got a build that runs `npm install -g serverless@2.x.x`. It's giving the same "Cannot run local installation of Serverless" message.

Log output: https://gist.github.com/StuartBertram/5325fc6dfe873ea18e8ca70d94d4bad0

We're not necessarily looking to update to 3.x right now. If there's a way to keep our existing 2.x build then that would be great :)

It's a Python lambda project, so there shouldn't be any rogue dependencies creeping in from other Node code.

**Reply by mnapoli** [reply]
@StuartBertram in your case it seems that the global v2 installation is trying to run v3 locally.

Do you have `serverless` v3 installed by `package.json` in your project? (if you have it in your project dependencies you will want to restrict it to `^2` only.

**Reply by StuartBertram** [reply]
Most of our `package.json` files look like:
```
{
  "name": "<project-name>",
  "description": "",
  "version": "0.1.0",
  "devDependencies": {
    "serverless-python-requirements": "^5.1.1"
  }
}
```

One of them also lists `serverless-api-gateway-throttling`, `serverless-domain-manager` and `serverless-prune-plugin`. `package.json` only exists for Serverless and we have `pyproject.toml`/`poetry.lock` for the code.

**Reply by StuartBertram** [reply]
I've put the steps from the failing job in a gist at https://gist.github.com/StuartBertram/d975cf5653165d8d8a525a09386a6c51. It fails at the "deploy" step.

**Reply by mnapoli** [reply]
Thanks for clarifying. It might be very likely that the reason for all of this is that the related plugins have `serverless` in their `peerDependencies` and they allow v3 (e.g. `^2 || ^3`).

That causes NPM 7 and up to install the latest available version of the peer dependency -> in other words that causes serverless v3 to be installed in `node_modules`.

That could explain the conflict with the global v2 version.

We're thinking about this but it would seem the most immediate solution for you would be to explicitly require `serverless` v2 in those projects? That way the local version will stick to v2, not v3.

Does that make sense?

**Reply by StuartBertram** [reply]
Yeah, that makes sense - although it is reporting v2, so I don't quite understand which part of NPM decided to also install v3 somewhere! I'll see what I can do 🙂

**Reply by Gkrlo** [reply]
Hi there! :) 
Maybe this can help someone... I was facing the same issue.I ended up installing the proper version locally and run all the commands with npx. 
e.g: 
- npm install serverless@2
- npx serverless plugin install -n serverless-python-requirements
- npx serverless deploy

**Reply by StuartBertram** [reply]
It looks like my issue is a combination of a global `npm install` of Serverless v2 and then a `serverless plugin install` that results in v3 being installed within the local environment. If I do a global install and then `npm ci` to install the plugins from the `package.json` then it seems to work. I can also do a local Serverless install and then a single `npm ci`, which adds lots of dependencies to my lockfile.

**Comment by PrimeObjects** [comment]
Can I use a local installed serverless to "serverless create" a project. because I want to keep the global version at 2.7 for other projects.
